fix(ivy): ngtsc is unable to detect flat module entry-point on windows

devversion opened the pull request:
* See individual commits

devversion committed:
refactor(ivy): fix incorrect error message in ngtsc "PathSegment"

gkalpak commented on packages/compiler-cli/src/ngtsc/shims/src/host.ts:
```diff
@@ -101,7 +104,7 @@ export class GeneratedShimsHostWrapper implements ts.CompilerHost {
     // 2) at least one of the shim generators recognizes it.
     return this.delegate.fileExists(fileName) ||
-        this.shimGenerators.some(gen => gen.recognize(fileName));
+        this.shimGenerators.some(gen => gen.recognize(AbsoluteFsPath.fromUnchecked(fileName)));
```
It might be worth adding a comment explaining why this is always expected to be absolute (similar to [this one][1]).

[1]: https://github.com/angular/angular/pull/29453/files?file-filters%5B%5D=.ts#diff-899e903e9872c26135ea51716d38e2b3R64

devversion replied:
Yeah let's add one here as well. I'm afraid we could technically also add comments to the generators as well, but that sounds like overkill :smile:

gkalpak commented on packages/compiler-cli/test/ngtsc/ngtsc_spec.ts:
```diff
@@ -2669,6 +2669,21 @@ describe('ngtsc behavioral tests', () => {
+          .toContain(
+              'export * from \'./index\';',
+              'Should detect the "index.ts" file as flat module entry-point.');
+    });
```
Does this test fail without the changes???

devversion replied:
It fails on Windows without my changes :smile: Though since we don't have any test for this logic at all, I figured I add one.

devversion committed:
fix(ivy): ngtsc is unable to detect flat module entry-point on windows

Currently when building an Angular project with `ngtsc`
and `flatModuleOutFile` enabled, the Ngtsc build will fail
if there are multiple source files as root file names.

Ngtsc and NGC currently determine the entry-point for multiple
root file names by looking for files ending with `/index.ts`.

This functionality is technically deprecated, but still supported
and currently breaks on Windows as the root file names are not
guaranteed to be normalized POSIX-like paths.

In order to make this logic more reliable in the future, this commit
also switches the shim generators and entry-point logic to the branded
path types. This ensures that we don't break this in the future.

devversion commented:
Caretaker note: Merge assistance because `test_docs_examples` fails due to a flake.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/shims/src/host.ts:
```diff
@@ -60,14 +61,16 @@ export class GeneratedShimsHostWrapper implements ts.CompilerHost {
       const generator = this.shimGenerators[i];
-      if (generator.recognize(fileName)) {
+      // TypeScript internal paths are guaranteed to be POSIX-like absolute file paths.
+      const absoluteFsPath = AbsoluteFsPath.fromUnchecked(fileName);
```
It turns out (at least in tests) that it is not true that `fileName` is always absolute. Changing this to `AbsoluteFsPath.from()` causes lots of test failures such as:

```
193) compiler compliance: listen() should create declare inputs/outputs
  Message:
    Error: Internal Error: AbsoluteFsPath.from(lib.d.ts): path is not absolute
  Stack:
    Error: Internal Error: AbsoluteFsPath.from(lib.d.ts): path is not absolute
        at Object.from (packages/compiler-cli/src/ngtsc/path/src/types.ts:59:13)
        at GeneratedShimsHostWrapper.getSourceFile (packages/compiler-cli/src/ngtsc/shims/src/host.ts:76:45)
```

@devversion & @alxhub
